src/main.rs

```diff
@@ -0,0 +1,57 @@
+use rand::Rng;
+use std::fmt;
+
+fn main() {
+    let mut bag = TileBag::new_default();
+    for i in 0..1000 {
+        let tile = bag.draw_tile();
+        println!("{}", tile)
+    }
+}
+
+struct TileBag {
+    bag: Vec<char>,
+    rng: rand::rngs::ThreadRng,
+}
+
+impl TileBag {
+    fn new(letter_distribution: [u8; 26]) -> Self {
+        let mut bag: Vec<char> = vec![];
+        let range: std::ops::Range<u8> = std::ops::Range { start: 0, end: 26 };
+        for letter in range {
+            for i in 0..letter_distribution[letter as usize] {
+                bag.push((letter + 65) as char);
+            }
+        }
```

Comment:
Generally I see people reach for iterators and maps here. So you could replace this range and the lines above with something like:

```rust
let bag: Vec<_> = letter_distribution.iter().enumerate()
  .map(|(letter, count)| (letter as u8, *count as usize))
  .flat_map(|(letter, count)| [((letter) + 65) as char].repeat(count))
  .collect();
```

`.enumerate()` turns an iterator over `T` into an iterator over `(index, T)` which saves us having to do the double loop.

The first `map` is a little dirty, but we have to do a bit of type wrangling and it's nice to encapsulate that into a new iterator over `(u8, usize)`

`flat_map` is super handy. There we return a vec/slice with the `repeat` function, which normally when collected would give you `Vec<Vec<_>>`. `flat_map` will flatten that down to a single `Vec<_>`. (`flat_map` also handles going from a `Vec<Option<T>>` to a `Vec<T>`, super handy)

We need the type definition of `let bag: Vec<_>` because `collect` needs to know what type to collect into — you could collect into a slice, or you can even collect chars into a String. We need to tell it to collect into a Vec, but we can elide the inner type with `_` and let Rust figure that bit out since it can be inferred.

Reply:
The compiler doesn't seem to require `Vec<_>`, did you add it due to convention?
Also, is `flat_map` essentially the same as chaining `map` then `flatten`?

Also, realised I don't really need the initial letter distribution values to be `u8`, I was thinking I needed to cast them to `char`s, but those values represent the count of each letter, not the letters themselves. So we can get rid of the `*count as usize` part, and remove the first `map`. Otoh, we are slightly less space efficient by storing 26 `usize`s rather than `u8`s - what do you think of the tradeoff?

Reply:
> The compiler doesn't seem to require Vec<_>, did you add it due to convention?

Huh, it can obviously figure this one out. I didn't test it without, in _most_ cases with `collect` you need to specify the type you're collecting into, so I often add that one out of habit. If it isn't needed, all the better!

>  Also, is flat_map essentially the same as chaining map then flatten?

✨ just _even more compact ✨

> What do you think of the tradeoff?

I'd say go with the easier-to-write option until it becomes a problem 😄 I use usize a lot for the convenience of indexing things even if it's far more than I'd need (until I'm doing something like storing/sending data over a network)